src/v/io/cache.h

```diff
@@ -0,0 +1,662 @@
+ *         cache_hook hook;
+ *     };
+ *
+ * The cache hook also provides an interface between the entry and the cache for
+ * communicating entry state changes such as the entry being evicted, or the
+ * application indicating that the entry has been used in a meaningful way.
+ *
+ * When the application using the cache rehydrates an entry on the ghost queue
+ * it is important to maintain the metadata contained in the cache hook. If the
+ * entry structure is re-used then it should be sufficient to avoid resetting
+ * the cache hook member. However, if the entry is destroyed and a replacement
+ * is created, then the application should initialize the hook in the new entry
+ * from the hook in the previous entry.
+ */
+class cache_hook {
+public:
+    cache_hook() noexcept = default;
+
+    /**
+     * Copy constructor.
+     *
+     * The value of evicted() will be false for the newly constructed hook.
+     */
+    cache_hook(const cache_hook&) = default;
+
+    cache_hook& operator=(const cache_hook&) = delete;
+    cache_hook(cache_hook&&) noexcept = delete;
+    cache_hook& operator=(cache_hook&&) noexcept = delete;
+    ~cache_hook() noexcept = default;
+
+    /**
+     * Mark a notable interaction with the entry containing this hook.
+     */
+    void touch() noexcept { freq_ = std::min(freq_ + 1, 3); }
+
+    /**
+     * Check if an entry containing this hook is in the cache.
+     *
+     * This method only considers the small and main cache queues, and does not
+     * reflect membership on the ghost queue. This method will return true for a
+     * default constructured hook.
+     *
+     * An entry need not check if an entry has been evicted using this
+     * interface, and may instead arrange for a side-effect of the cache's
+     * configured eviction function to mark an entry as evicted. Consider for
+     * example the following entry and eviction function:
+     *
+     *     struct entry { std::unique_ptr<int> ptr; }
+     *
+     *     struct evict {
+     *         bool operator()(entry& e) {
+     *             e.ptr = nullptr;
+     *             return true;
+     *         }
+     *     };
+     *
+     * An application may then use `e.ptr` to indicate that a entry has been
+     * evicted instead of invoking cache_hook::evicted.
+     */
+    [[nodiscard]] bool evicted() const noexcept { return !hook_.is_linked(); }
+
+private:
+    using hook_type = boost::intrusive::list_member_hook<
+      boost::intrusive::link_mode<boost::intrusive::safe_link>>;
+
+    /*
+     * the location is checked only in cache::erase where the case of the hook
+     * being on neither main nor small is indicated by evicted() being true.
+     */
+    enum class location : uint8_t { main, small };
+
+    template<
+      typename T,
+      cache_hook T::*Hook,
+      cache_evictor<T> Evictor,
+      cache_cost<T> Cost>
+    friend class cache;
+
+    friend class testing_details::cache_hook_accessor;
+
+    uint8_t freq_{0};
```

Comment:
> Certain sets of names and function signatures are always reserved to the implementation:
> 
> [(1.1)](https://timsong-cpp.github.io/cppwp/n3337/reserved.names#global.names-1.1)
> Each name that contains a double underscore __ or begins with an underscore followed by an uppercase letter ([[lex.key]](https://timsong-cpp.github.io/cppwp/n3337/lex.key)) is reserved to the implementation for any use.
> 
> [(1.2)](https://timsong-cpp.github.io/cppwp/n3337/reserved.names#global.names-1.2)
> Each name that begins with an underscore is reserved to the implementation for use as a name in the global namespace.


Technically, it's not correct to use leading underscore but our whole codebase if littered with it.

Reply:
It doesn’t seem that we are breaking the rules. As you quoted, they are reserved only in the global scope. It can lead to some surprises for sure :) 

https://devblogs.microsoft.com/oldnewthing/20230109-00/?p=107685

Reply:
🌶️ Time to adopt Google Style guide for naming? 🌶️ 

https://google.github.io/styleguide/cppguide.html#Variable_Names

Reply:
at least the good parts